### ISSUE TITLE
Base types give incorrect type names

### ISSUE THREAD

**Issue by michaelcheers** [bridgedotnet/Bridge#1347]
**UPD** Rewriting the issue description according to the following:

### These items have been fixed earlier
- Array (s) don't give types and give type 'Array'. (fixed by #2056)
- Type (s) give Bridge.{type} not System.{type} (eg. Bridge.Int32 not System.Int32)

### These items will be fixed under #2077 (PR #2097)
- Double (s) which can be int (s) are int (s) with GetType eg(3.0)
- Float (s) give type double
- Built in type (s) don't give mscorlib (in part of List etc)
- Some type (s) give Int32 or Double as a type not their type (eg. byte, uint, ushort, short, int, any enum)

### These items are considered as designed
- Anonymous type (s) give $AnonymousType$(Number of fields) not mscorlib type

A [Deck](http://deck.net/677376af262b962ff43c8ea6ae045346) sample for all the passing items **above**.

A [Deck](http://deck.net/cfc191b9a0c4f04dbcd14a2a17974da6) sample for all the currently failing items **below**.

### These items have already issues created
- Function types (eg. Action and Func) give the type Function (details added into #1335)
- Type (s) give type 'Function' not System.RuntimeType (details added into #1335)
- Enum type name formatting if nested in a class #2062

### New issues created (which are not related to each other)
- System.String #2307
- System.Boolean #2308

## Initial

This is a combination of lots of type issues and a few new ones:
- Function types (eg. Action and Func) give the type Function (issue #1335)
- Double (s) which can be int (s) are int (s) with GetType eg(3.0) (fixed by #2077)
- Float (s) give type double (fixed by #2077)
- Array (s) don't give types and give type 'Array'. (fixed by #2056)
- Type (s) give type 'Function' not System.RuntimeType (issue #1335)
- Type (s) give Bridge.{type} not System.{type} (eg. Bridge.Int32 not System.Int32). (fixed by #1337)
- Built in type (s) don't give mscorlib
- Some type (s) give Int32 or Double as a type not their type (eg. byte, uint, ushort, short, int, any enum) (fixed by #2077)
- Anonymous type (s) give $AnonymousType$(Number of fields) not mscorlib type
### Expected

``` javascript
System.String, System.Int32, System.Byte, System.UInt32, System.Int64, System.UInt64, System.Int16, System.UInt16, System.SByte, System.String[], System.Func`2[[System.Int32, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089],[System.String, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]], System.RuntimeType, System.Boolean, App+Enum, System.RuntimeType, System.RuntimeType, System.RuntimeType, System.RuntimeType, System.Double, System.Single, System.Char, System.Decimal, <>f__AnonymousType0`1[[System.Int32, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]], System.Collections.Generic.List`1[[System.Int32, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]]
```
### Actual

``` javascript
String, Bridge.Int32, Bridge.Int32, Bridge.Int32, Bridge.Long, Bridge.ULong, Bridge.Int32, Bridge.Int32, Bridge.Int32, Array, Function, Function, Boolean, Bridge.Int32, Function, Function, Function, Function, Bridge.Int32, Bridge.Int32, Bridge.Int32, Bridge.Decimal, Object, Bridge.List$1$Bridge.Int32
```
### Steps To Reproduce

[Dot Net Fiddle](https://dotnetfiddle.net/DUwfW1)
[Deck](http://deck.net/0dbba45615b0140966df7e95d57f70c8)

``` csharp
public class App
{
    public static void Main()
    {
        var array = new object[]
        {
            "abc",
            7,
            (byte)7,
            7u,
            7l,
            7lu,
            (short)7,
            (ushort)8,
            (sbyte)9,
            new[]{"Hello World"},
            ((Func<int, string>)(v => "Hello World")),
            7.GetType(),
            false,
            Enum.Value,
            //typeof(void),
            typeof(Type),
            typeof(Action),
            typeof(Func<int>),
            typeof(Enum),
            3.0,
            3f,
            'a',
            7m,
            new{Id=3}
        };
        
        Console.WriteLine(string.Join(", ", array.ToList().ConvertAll(v => v.GetType().FullName).ToArray()));
    }

    enum Enum
    {
        Value
    }
}
```

**Comment by LeonidVeriga** [comment]
Related: #1294 #1293 #1292 #1248

**Comment by michaelcheers** [comment]
JSIL has the same problem: [JSIL issue](https://github.com/sq/JSIL/issues/987).

**Comment by geoffreymcgill** [comment]
Adding this to Bridge 16.0 milestone. We need to fix this.

**Comment by Porl91** [comment]
@geoffreymcgill Hi Geoffrey, I just wanted to check whether there's a planned fix for point 4 in this issue regarding arrays being non-typed? I'm currently running into some trouble when attempting to serialise / deserialise strongly typed arrays.

**Comment by geoffreymcgill** [comment]
HI @Porl91 - Revising the types is scheduled for Bridge 16.0. We haven't completed this work yet, but it is one of the highest priority items for Bridge 16 and we will not release 16 until this has been fixed. 

At the moment, there is no firm date for the release of Bridge 16, but there is a [v16](http://new.bridge.net/pricing) branch if you want to build locally and follow along with the progress.

**Comment by Porl91** [comment]
Hi @geoffreymcgill, thanks for replying. That's good news!

**Comment by michaelcheers** [comment]
@geoffreymcgill, this is probably going to break a lot of Template functions using object.

**Comment by LeonidVeriga** [comment]
Hi @Porl91, the point 4 mentioned in your https://github.com/bridgedotnet/Bridge/issues/1347#issuecomment-263328892 has been fixed by #2056 for **15.7** release.

**Comment by LeonidVeriga** [comment]
Hi @michaelcheers, yes we had to change some Templates.

**Comment by LeonidVeriga** [comment]
Rewriting the issue description according to the followng

### These items have been fixed earlier
- Array (s) don't give types and give type 'Array'. (fixed by #2056)
- Type (s) give Bridge.{type} not System.{type} (eg. Bridge.Int32 not System.Int32)

### These items will be fixed under #2077 (PR #2097)
- Double (s) which can be int (s) are int (s) with GetType eg(3.0)
- Float (s) give type double
- Built in type (s) don't give mscorlib (in part of List etc)
- Some type (s) give Int32 or Double as a type not their type (eg. byte, uint, ushort, short, int, any enum)

### These items are considered as designed
- Anonymous type (s) give $AnonymousType$(Number of fields) not mscorlib type

A [Deck](http://deck.net/677376af262b962ff43c8ea6ae045346) sample for all the passing items **above**.


A [Deck](http://deck.net/cfc191b9a0c4f04dbcd14a2a17974da6) sample for all the currently failing items **below**.

### These items have already issues created
- Function types (eg. Action and Func) give the type Function (details added into #1335)
- Type (s) give type 'Function' not System.RuntimeType (details added into #1335)
- Enum type name formatting if nested in a class #2062

### New issues created (which are not related to each other)
- System.String #2307
- System.Boolean #2308
